File: services/menos-import/api/menos/auth/keys.py
"""SSH public key management."""

import hashlib
import re
from base64 import b64decode
from pathlib import Path

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
from cryptography.hazmat.primitives.serialization import load_ssh_public_key


class KeyStore:
    """Manages authorized SSH public keys."""

    def __init__(self, keys_path: Path):
        self.keys_path = keys_path
        self._keys: dict[str, Ed25519PublicKey] = {}
        self._load_keys()

    def _load_keys(self) -> None:
        """Load all authorized keys from disk."""
        if not self.keys_path.exists():
            return

        # Support both single file (authorized_keys) and directory of keys
        if self.keys_path.is_file():
            self._load_authorized_keys_file(self.keys_path)
        elif self.keys_path.is_dir():
            # Load authorized_keys if exists
            auth_keys = self.keys_path / "authorized_keys"
            if auth_keys.exists():
                self._load_authorized_keys_file(auth_keys)

            # Load individual .pub files
            for pub_file in self.keys_path.glob("*.pub"):
                self._load_key_file(pub_file)

    def _load_authorized_keys_file(self, path: Path) -> None:
        """Load keys from authorized_keys format file."""
        for line in path.read_text().splitlines():
            line = line.strip()
            if not line or line.startswith("#"):
                continue
            self._parse_and_store_key(line)

    def _load_key_file(self, path: Path) -> None:
        """Load a single public key file."""
        content = path.read_text().strip()
        self._parse_and_store_key(content)

    def _parse_and_store_key(self, key_line: str) -> None:
        """Parse SSH public key line and store if ed25519."""
        # Format: ssh-ed25519 AAAA... comment
        match = re.match(r"^(ssh-ed25519)\s+(\S+)(?:\s+(.*))?$", key_line)
        if not match:
            return

        key_type, key_data, comment = match.groups()

        try:
            key = load_ssh_public_key(key_line.encode())
            if isinstance(key, Ed25519PublicKey):
                key_id = self._compute_key_id(key_data)
                self._keys[key_id] = key
        except Exception:
            pass  # Skip invalid keys

    def _compute_key_id(self, key_data: str) -> str:
        """Compute key ID (fingerprint) from base64 key data."""
        raw = b64decode(key_data)
        digest = hashlib.sha256(raw).hexdigest()
        return f"SHA256:{digest[:16]}"

    def get_key(self, key_id: str) -> Ed25519PublicKey | None:
        """Get public key by ID."""
        return self._keys.get(key_id)

    def list_key_ids(self) -> list[str]:
        """List all authorized key IDs."""
        return list(self._keys.keys())

    def reload(self) -> None:
        """Reload keys from disk."""
        self._keys.clear()
        self._load_keys()

File: services/core/src/vault/keys.ts
import { createHash } from "node:crypto";
import { readFileSync, readdirSync, statSync, existsSync } from "node:fs";
import { join } from "node:path";

const SSH_ED25519 = "ssh-ed25519";

export type AuthorizedKey = {
  keyId: string;
  publicKeyRaw: Buffer;
  comment: string | undefined;
};

function extractRawEd25519(blob: Buffer): Buffer | undefined {
  if (blob.length < 4) return undefined;
  const typeLength = blob.readUInt32BE(0);
  if (blob.length < 4 + typeLength + 4) return undefined;
  const keyType = blob.subarray(4, 4 + typeLength).toString("utf8");
  if (keyType !== SSH_ED25519) return undefined;
  const keyLength = blob.readUInt32BE(4 + typeLength);
  const start = 4 + typeLength + 4;
  if (keyLength !== 32 || blob.length < start + keyLength) return undefined;
  return blob.subarray(start, start + keyLength);
}

export function computeKeyId(blob: Buffer): string {
  const digest = createHash("sha256").update(blob).digest("hex");
  return `SHA256:${digest.slice(0, 16)}`;
}

export function parseAuthorizedKeyLine(line: string): AuthorizedKey | undefined {
  const trimmed = line.trim();
  if (trimmed === "" || trimmed.startsWith("#")) return undefined;
  const match = /^ssh-ed25519\s+(\S+)(?:\s+(.*))?$/.exec(trimmed);
  if (!match) return undefined;
  const keyData = match[1];
  if (keyData === undefined) return undefined;
  if (!/^[A-Za-z0-9+/=]+$/.test(keyData)) return undefined;
  const blob = Buffer.from(keyData, "base64");
  const raw = extractRawEd25519(blob);
  if (raw === undefined) return undefined;
  return { keyId: computeKeyId(blob), publicKeyRaw: raw, comment: match[2] };
}

/**
 * Manages authorized SSH public keys.
 *
 * Parity contract: accepts an authorized_keys file or a directory containing
 * an authorized_keys file plus individual `*.pub` files. Only ed25519 keys
 * are stored. Key ids are `SHA256:` plus the first 16 hex characters of the
 * SHA-256 digest of the SSH wire-format key blob, matching both the Python
 * server and the unmodified dotfiles client signer.
 */
export class KeyStore {
  private readonly keysPath: string;
  private keys = new Map<string, Buffer>();

  constructor(keysPath: string) {
    this.keysPath = keysPath;
    this.loadKeys();
  }

  private loadKeys(): void {
    if (!existsSync(this.keysPath)) return;
    const stats = statSync(this.keysPath);
    if (stats.isFile()) {
      this.loadAuthorizedKeysFile(this.keysPath);
      return;
    }
    if (!stats.isDirectory()) return;
    const authKeys = join(this.keysPath, "authorized_keys");
    if (existsSync(authKeys)) {
      this.loadAuthorizedKeysFile(authKeys);
    }
    for (const entry of readdirSync(this.keysPath)) {
      if (!entry.endsWith(".pub")) continue;
      this.storeLine(readFileSync(join(this.keysPath, entry), "utf8").trim());
    }
  }

  private loadAuthorizedKeysFile(path: string): void {
    for (const line of readFileSync(path, "utf8").split(/\r?\n/)) {
      this.storeLine(line);
    }
  }

  private storeLine(line: string): void {
    const parsed = parseAuthorizedKeyLine(line);
    if (parsed !== undefined) {
      this.keys.set(parsed.keyId, parsed.publicKeyRaw);
    }
  }

  getKey(keyId: string): Buffer | undefined {
    return this.keys.get(keyId);
  }

  listKeyIds(): string[] {
    return [...this.keys.keys()];
  }

  reload(): void {
    this.keys.clear();
    this.loadKeys();
  }
}
